daemons/gptp/common/ieee1588port.cpp

```diff
@@ -477,427 +568,528 @@ void IEEE1588Port::processEvent(Event e)
 
 		clock->putTimerQLock();
 
-		break;
-	case STATE_CHANGE_EVENT:
-		if ( clock->getPriority1() != 255 ) {
-			int number_ports, j;
-			PTPMessageAnnounce *EBest = NULL;
-			char EBestClockIdentity[PTP_CLOCK_IDENTITY_LENGTH];
-
-			IEEE1588Port **ports;
-			clock->getPortList(number_ports, ports);
-
-			/* Find EBest for all ports */
-			j = 0;
-			for (int i = 0; i < number_ports; ++i) {
-				while (ports[j] == NULL)
-					++j;
-				if (ports[j]->port_state == PTP_DISABLED
-				    || ports[j]->port_state == PTP_FAULTY) {
-					continue;
+		if (automotive_profile) {
+			if (testMode) {
+				stationState = STATION_STATE_ETHERNET_READY;
+				APMessageTestStatus *testStatusMsg = new APMessageTestStatus(this);
+				if (testStatusMsg) {
+					testStatusMsg->sendPort(this);
+					delete testStatusMsg;
 				}
-				if (EBest == NULL) {
-					EBest = ports[j]->calculateERBest();
-				} else {
-					if (ports[j]->calculateERBest()->isBetterThan(EBest)) {
-						EBest = ports[j]->calculateERBest();
-					}
+			}
+			if (!isGM) {
+				// Send an initial signalling message
+				PTPMessageSignalling *sigMsg = new PTPMessageSignalling(this);
+				if (sigMsg) {
+					sigMsg->setintervals(log_min_mean_pdelay_req_interval, log_mean_sync_interval, PTPMessageSignalling::sigMsgInterval_NoSend);
+					sigMsg->sendPort(this, NULL);
+					delete sigMsg;
 				}
 			}
+		}
 
-			/* Check if we've changed */
-			{
+		break;
 
-			  uint8_t LastEBestClockIdentity[PTP_CLOCK_IDENTITY_LENGTH];
-			  clock->getLastEBestIdentity().
-				  getIdentityString( LastEBestClockIdentity );
-			  EBest->getGrandmasterIdentity( EBestClockIdentity );
-			  if( memcmp
-				  ( EBestClockIdentity, LastEBestClockIdentity,
-					PTP_CLOCK_IDENTITY_LENGTH ) != 0 )
-			  {
-				  ClockIdentity newGM;
-				  changed_external_master = true;
-				  newGM.set((uint8_t *) EBestClockIdentity );
-				  clock->setLastEBestIdentity( newGM );
-			  } else {
-				  changed_external_master = false;
-			  }
-			}
+	case STATE_CHANGE_EVENT:
+		if (!automotive_profile) {       // BMCA is not active with Automotive Profile
+			if ( clock->getPriority1() != 255 ) {
+				int number_ports, j;
+				PTPMessageAnnounce *EBest = NULL;
+				char EBestClockIdentity[PTP_CLOCK_IDENTITY_LENGTH];
+
+				IEEE1588Port **ports;
+				clock->getPortList(number_ports, ports);
+
+				/* Find EBest for all ports */
+				j = 0;
+				for (int i = 0; i < number_ports; ++i) {
+					while (ports[j] == NULL)
```

Comment:
does getPortList guarantee that ports is always a valid pointer and that number_ports is always within the range of the pointer? (there is no error check on return of clock->getPortList)